File: BasicDataStructure/Hash/longestConsecutive.py
from typing import List

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # 1.用哈希集合查找 2.去重
        s = set(nums)
        best = 0

        for x in s:
            # 只从“序列起点”开始扩展，避免重复计算
            # 一个数x是某个连续序列的起点当且仅当x-1不在集合中
            if x - 1 not in s:
                cur = x
                length = 1
                # 从起点向右扩展，计算长度
                while cur + 1 in s:
                    cur += 1
                    length += 1
                best = max(best, length)

        return best
"""
题目要求：
    找 数值连续 的最长序列
    不要求在原数组中连续
    要求 O(n) 时间复杂度


思路：
    利用哈希集合O(1)查找，只从连续序列起点开始扩展，每个元素最多访问一次，因此时间复杂度O(n), 空间复杂度O(n)

s = set(nums)的作用：

    

"""

File: BasicDataStructure/Hash/longestConsecutive.cpp
#include <vector>
#include <string>
#include <unordered_set>
#include <algorithm>
using namespace std;

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        // 用哈希集合存储所有数字，方便 O(1) 判断某个数字是否存在
        unordered_set<int> st(nums.begin(), nums.end());

        int ans = 0;  // 记录最长连续序列的长度

        // 遍历集合中的每个数字
        for (int x : st) {
            // 如果 x-1 存在，说明 x 不是连续序列的起点
            // 只有当 x-1 不存在时，x 才是起点
            if (st.count(x - 1)) {
                continue;
            }

            // 从 x 开始向后寻找连续的数字
            int cur = x;
            int len = 1;

            while (st.count(cur + 1)) {
                cur++;
                len++;
            }

            // 更新答案
            ans = max(ans, len);
        }

        return ans;
    }
};